pkgs/by-name/ze/zen-browser-unwrapped/package.nix

```diff
@@ -0,0 +1,436 @@
+    "--with-system-webp"
+    "--with-system-zlib"
+    "--with-wasi-sysroot=${wasiSysRoot}"
+    "--host=${buildStdenv.buildPlatform.config}"
+    "--target=${buildStdenv.hostPlatform.config}"
+    (lib.enableFeature alsaSupport "alsa")
+    (lib.enableFeature ffmpegSupport "ffmpeg")
+    (lib.enableFeature geolocationSupport "necko-wifi")
+    (lib.enableFeature gssSupport "negotiateauth")
+    (lib.enableFeature jackSupport "jack")
+    (lib.enableFeature jemallocSupport "jemalloc")
+    (lib.enableFeature pulseaudioSupport "pulseaudio")
+    (lib.enableFeature sndioSupport "sndio")
+    (lib.enableFeature webrtcSupport "webrtc")
+    # --enable-release adds -ffunction-sections & LTO that require a big amount
+    # of RAM, and the 32-bit memory space cannot handle that linking
+    (lib.enableFeature (!debugBuild && !stdenv.hostPlatform.is32bit) "release")
+    (lib.enableFeature enableDebugSymbols "debug-symbols")
+  ] ++ lib.optional stdenv.hostPlatform.isAarch "--disable-wasm-avx";
+
+  configureScript = writeShellScript "configureMozconfig" (
+    (lib.optionalString stdenv.hostPlatform.isAarch ''
+      echo "ac_add_options --with-libclang-path=/usr/lib64" >> ./configs/linux/mozconfig
+
+      # linux mozconfig
+      sed -i 's/x86-\(64\|64-v3\)/native/g' ./configs/linux/mozconfig
+      sed -i 's/x86_64-pc-linux/aarch64-linux-gnu/g' ./configs/linux/mozconfig
+
+      # eme/widevine must be disabled on arm64 (thx google)
+      sed -i '/--enable-eme/s/^/# /' ./configs/common/mozconfig
+      sed -i 's/-msse3//g' ./configs/linux/mozconfig
+      sed -i 's/-mssse3//g' ./configs/linux/mozconfig
+      sed -i 's/-msse4.1//g' ./configs/linux/mozconfig
+      sed -i 's/-msse4.2//g' ./configs/linux/mozconfig
+      sed -i 's/-mavx2//g' ./configs/linux/mozconfig
+      sed -i 's/-mavx//g' ./configs/linux/mozconfig
+      sed -i 's/-mfma//g' ./configs/linux/mozconfig
+      sed -i 's/-maes//g' ./configs/linux/mozconfig
+      sed -i 's/-mpopcnt//g' ./configs/linux/mozconfig
+      sed -i 's/-mpclmul//g' ./configs/linux/mozconfig
+      sed -i 's/+avx2//g' ./configs/linux/mozconfig
+      sed -i 's/+sse4.1//g' ./configs/linux/mozconfig
+
+    '')
+    + ''
+      for flag in $@; do
+        echo "ac_add_options $flag" >> mozconfig
+      done
+    ''
+  );
+
+  # To the person reading this wondering what is going on here, this is what
+  # happens when a build process relies on Git. Normally you would use `fetchgit`
+  # with `leaveDotGit = true`, however that leads to reproducibility issues, so
+  # instead we create our own Git repo with a single commit.
+  #
+  # `surfer` (the build tool made for zen-browser) uses git to read the latest
+  # HEAD commit, `git apply`, and likely a few other operations.
+  preConfigure = ''
+    export HOME="$TMPDIR"
+    git config --global user.email "nixbld@localhost"
+    git config --global user.name "nixbld"
+    git init
+    git add --all
+    git commit -m 'nixpkgs'
```

Comment:
Isn't this not reproducible as well? git commit will create a commit with different hash everytime.

Reply:
Potentially, but if that was the case we likely would have seen multiple hash mismatches by now. It can only cause reproducibility issues if the commit hash is embedded in the end result.

Reply:
I am wondering if we can do away with a fake git command, that returns the version (just a dummy shell script). Why do we need git here in the first place?

Reply:
There is a comment explaining why, `surfer` uses multiple git commands, including reading the commit hash (which may be a problem) and `git apply`.

The only way around this is to patch `surfer` to get rid of all usage of `git` commands. While I hate this workaround, unless it causes reproducibility issues, I don't see how maintaining patches for the build tool is any better.

Reply:
I am sorry if that question sounds dumb, but why is the following statement true?

> Normally you would use `fetchgit` with `leaveDotGit = true`, however that leads to reproducibility issues, ..."

Reply:
> Isn't this not reproducible as well? git commit will create a commit with different hash everytime.

I might be wrong here, but I think git hashes are deterministic. They're derived from the actual committed changes. Of course, there might be factors at play here that could still cause non-reproducable results.

Reply:
Sure, they're deterministic -- but I'm pretty sure they are also derived from the commit time.

Reply:
After doing a little research it seems the hashed data is supposed to include the system time down to the second. 

- [Why does git commit --amend change the hash even if I don't make any changes?](https://stackoverflow.com/questions/23791999/why-does-git-commit-amend-change-the-hash-even-if-i-dont-make-any-changes)
- [How is git commit sha1 formed](https://gist.github.com/masak/2415865)
- [What are Git hashes?](https://graphite.dev/guides/git-hashing#how-is-a-git-hash-generated)

Reply:
If the only thing that changes the commit hashes besides changes are the time that they are commited, these environment variables `GIT_AUTHOR_DATE` and `GIT_COMMITTER_DATE` could be set to a specific time, so they are the same every build.

Reply:
> I am sorry if that question sounds dumb, but why is the following statement true?
> 
> > Normally you would use `fetchgit` with `leaveDotGit = true`, however that leads to reproducibility issues, ..."

The issue is the contents of the `.git` folder may not always be exactly bit-to-bit reproducible. If we use `leaveDotGit = true` the `src` derivation will contain a `.git` folder which is then placed into the Nix store. So if that `.git` folder at all changes (which has happened previously while making this package) the `src` derivation becomes unreproducible.

Instead, if we make `.git` folder during the build that doesn't get copied to the Nix store, you avoid the reproducibility issues of the `.git` folder itself being different, even if the commit hash, git log, etc all are identical.

I do realize how strange this sounds, but since the change was made to remove `leaveDotGit` from this package, nobody has reported any reproducibility issues for this package.

Reply:
> This means that we will have to create a new `.git` every time Zen is updated, no?

Not sure what you mean. The `.git` folder is only created so `surfer` can run it's git commands. As long as it doesn't embed the commit hash or dates, there is no problem.

Zen updating doesn't change the package build other than the hash for `src` and `firefoxSrc` (if the Firefox version has changed).

Reply:
> these environment variables `GIT_AUTHOR_DATE` and `GIT_COMMITTER_DATE` could be set to a specific time, so they are the same every build.

Exactly, I am worried that the drv (not the src hash) might change with each build, even if they are built on the same zen version/commit. Can we somehow make it produce the same git hash everytime when committing?

Reply:
> So if that `.git` folder at all changes (which has happened previously while making this package) the `src` derivation becomes unreproducible.

This is a quite surprising result, after all `leaveDotGit`does some[ post-processing](https://github.com/NixOS/nixpkgs/blob/b4928123e82958b821bf652b27c3c87b80af5bcf/pkgs/build-support/fetchgit/nix-prefetch-git#L255) to avoid that. The `.git` folder should not change in a way that does not get cleaned up this way. I don't mean to suggest making this a stopper for merging, but it would be nice to investigate why this is happening and make a new issue for that.

Meanwhile, this workaround can be left here with a note to change it after the issue is fixed. Should be better than relying on a hack permanently.